### PR TITLE
#22 EA-2043 updated import scripts using wikidata pagerank index

### PR DESCRIPTION
**Pull request by rgraf**
Implemented querying of the new Solr

### CODE REVIEW

File: entity_collection/munge/mongo_import/config/harvester.properties.template [hunk 1]
```diff
@@ -16,4 +16,10 @@ organizations.harvester.mongo.port = -1
...
 harvester.relevance.solr.core.uri = http://localhost:9191/solr/search_production_publish_1/select?indent=on&wt=json
-harvester.relevance.ranking.model = normalized
\ No newline at end of file
+harvester.relevance.ranking.model = normalized
+
+#pagerank solr 
+pagerank.solr = 
+
+#search api solr
+search.api.solr =
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/config/harvester.properties.template]: what is this property used for? Is it different than the harvester.relevance.solr.core.uri?

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py [hunk 2]
```diff
@@ -224,13 +225,13 @@ def extract_numeric_id(self, entity_id):
         #numeric id is the last part of the URL 
         return parts[len(parts) - 1]
     
-    def build_solr_doc(self, entities, start, one_entity = False):
+    def build_solr_doc(self, entities, start, one_entity = False, importer=None):
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: why do you need the importer param? Isn't it a class attribute?

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py [hunk 3]
```diff
@@ -650,15 +654,18 @@ class IndividualEntityBuilder:
 
     def build_individual_entity(self, entity_id):
         from pymongo import MongoClient
-        import shutil
         if(entity_id.find("/place/") > 0):
             harvester = PlaceHarvester()
+            importer = MetricsImporter(harvester, MetricsImporter.DB_PLACE, MetricsImporter.TYPE_PLACE)
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: The IndividualEntityBuilder is used only for testing.  The importer must be initialized in the init method of each Harvester subclasses

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py [hunk 4]
```diff
@@ -276,11 +277,14 @@ def get_writepath(self, start, one_entity):
         else:
             return self.write_dir + "/" + self.name + "/" + self.name + "_" + str(start) + "_" + str(start + ContextClassHarvester.CHUNK_SIZE) +  ".xml"
 
-    def grab_relevance_ratings(self, docroot, entity_id, entity):
-        hitcounts = self.relevance_counter.get_raw_relevance_metrics(entity_id, entity)
-        eu_enrichments = hitcounts["europeana_enrichment_hits"]
-        eu_terms = hitcounts["europeana_string_hits"]
-        pagerank = hitcounts["pagerank"]
+    def grab_relevance_ratings(self, docroot, entity_id, entity, importer):
+        hitcounts = self.relevance_counter.get_raw_relevance_metrics(entity_id, entity, importer)
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: please rename variable name to metricsRecord

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py [hunk 5]
```diff
@@ -276,11 +277,14 @@ def get_writepath(self, start, one_entity):
         else:
             return self.write_dir + "/" + self.name + "/" + self.name + "_" + str(start) + "_" + str(start + ContextClassHarvester.CHUNK_SIZE) +  ".xml"
 
-    def grab_relevance_ratings(self, docroot, entity_id, entity):
-        hitcounts = self.relevance_counter.get_raw_relevance_metrics(entity_id, entity)
-        eu_enrichments = hitcounts["europeana_enrichment_hits"]
-        eu_terms = hitcounts["europeana_string_hits"]
-        pagerank = hitcounts["pagerank"]
+    def grab_relevance_ratings(self, docroot, entity_id, entity, importer):
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: please remove importer param and use it as class attribute

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py [hunk 6]
```diff
@@ -334,7 +338,7 @@ def process_created_modified_timestamps(self, docroot, entity_rows):
         if "modified" in entity_rows:
             self.add_field(docroot, "modified", entity_rows["modified"].isoformat()+"Z")
 
-    def process_representation(self, docroot, entity_id, entity):
+    def process_representation(self, docroot, entity_id, entity, importer):
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: please remove importer param and use it as class attribute

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py [hunk 7]
```diff
@@ -542,7 +546,7 @@ def build_entity_chunk(self, start):
             agents_chunk[agent_id] = self.client.annocultor_db.TermList.find_one({ 'codeUri' : agent_id })
         return agents_chunk
 
-    def build_entity_doc(self, docroot, entity_id, entity_rows):
+    def build_entity_doc(self, docroot, entity_id, entity_rows, importer):
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: please remove importer param and use it as class attribute

File: entity_collection/munge/mongo_import/entities/DepictionManager.py [hunk 8]
```diff
@@ -11,6 +11,7 @@ def load_depictions(self):
 		from urllib.parse import unquote
 
 		current_dir=os.getcwd()
+		current_dir = current_dir.replace("\\tests","")
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/DepictionManager.py]: please remove this line of code and set your run configurations to use mongo_import  folder as working directory

File: entity_collection/munge/mongo_import/entities/ranking_metrics/MetricsImporter.py [hunk 9]
```diff
@@ -75,6 +77,24 @@ def import_metrics(self):
         self.store_metrics()
                         
     
+    def import_metrics_for_entity(self, entity_id, wkdt_uri):
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/MetricsImporter.py]: please remove this method and integrate it into the import_metrics method, which must take an entity as optional parameter

File: entity_collection/munge/mongo_import/entities/ranking_metrics/MetricsImporter.py [hunk 10]
```diff
@@ -192,9 +212,9 @@ def fetch_metrics(self, entity):
         record.wikidata_id = self.harvester.relevance_counter.extract_wikidata_uri(entity)
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/MetricsImporter.py]: please move the extract_wikidata_uri to this class, it should not be used anymore in RelevanceCounter

File: entity_collection/munge/mongo_import/entities/ranking_metrics/MetricsImporter.py [hunk 11]
```diff
@@ -275,20 +295,29 @@ def get_page_rank(self, wikidata_id, all_pageranks):
...
         # process pagerank file, grab relevant items
         self.wikidata_pr_file =  os.path.join(os.path.dirname(__file__), self.WKDT_PAGE_RANK)
         with open(self.wikidata_pr_file) as ult:
             for line in ult.readlines():
                 (wkd_dbp_uri, pr) = line.split("\t")
                 
                 wkdt_uri = wkd_dbp_uri.replace(
-                    RelevanceCounter.WIKIDATA_DBPEDIA_PREFIX, 
-                    RelevanceCounter.WIKIDATA_PREFFIX)
+                    MetricsRecord().WIKIDATA_DBPEDIA_PREFIX,
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/MetricsImporter.py]: use constant and static referenciation, do no create new MetricsRecord objects for this

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py [hunk 12]
```diff
@@ -88,33 +60,35 @@ def db_connect(self):
             self.dbpath = os.path.join(os.path.dirname(__file__), 'db', self.name + ".db")
             self.db = slt.connect(self.dbpath)
         
-    def get_raw_relevance_metrics(self, uri, entity):
-        self.db_connect()
-        csr = self.db.cursor()
-        csr.execute("SELECT id, wikidata_id, wikipedia_hits, europeana_enrichment_hits, europeana_string_hits, pagerank FROM hits WHERE id='"+ uri + "'")
-        first_row = csr.fetchone()
+    def get_raw_relevance_metrics(self, uri, entity, importer):
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py]: please remove the importer parameter from the method and set it as class attribute within the constructor (init method)

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py [hunk 13]
```diff
@@ -88,33 +60,35 @@ def db_connect(self):
...
+        first_row = None
+        try:
+            self.db_connect()
+            csr = self.db.cursor()
+            csr.execute("SELECT id, wikidata_id, wikipedia_hits, europeana_enrichment_hits, europeana_string_hits, pagerank FROM hits WHERE id='"+ uri + "'")
+            first_row = csr.fetchone()
+        except:
+            # database is empty - calculate metrics
+            wikidata_id = self.extract_wikidata_uri(entity)
+            metrics = importer.import_metrics_for_entity(uri, wikidata_id)
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py]: please change the implementation and send the entity object as parameter to import_metrics method

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py [hunk 14]
```diff
@@ -88,33 +60,35 @@ def db_connect(self):
...
-        first_row = csr.fetchone()
+    def get_raw_relevance_metrics(self, uri, entity, importer):
+        first_row = None
+        try:
+            self.db_connect()
+            csr = self.db.cursor()
+            csr.execute("SELECT id, wikidata_id, wikipedia_hits, europeana_enrichment_hits, europeana_string_hits, pagerank FROM hits WHERE id='"+ uri + "'")
+            first_row = csr.fetchone()
+        except:
+            # database is empty - calculate metrics
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py:68]: According to the previous implementation, if the database is empty the first_row will be None and no exception will be raised. Here we can get connection or sql errors, but not an exception if the row is not found. If the database is initialized properly and the connection is successful, no exeption will be raised here

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py [hunk 15]
```diff
@@ -88,33 +60,35 @@ def db_connect(self):
...
+            wikidata_id = self.extract_wikidata_uri(entity)
+            metrics = importer.import_metrics_for_entity(uri, wikidata_id)
+            return metrics
+        
         if(first_row is not None):
             (_, wikidata_id, wikipedia_hits, europeana_enrichment_hits, europeana_string_hits, pagerank) = first_row
             if(pagerank is None):
                 pagerank = 0
-        else:
+        else:
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py:75]: please replace the implementation of the else block with importer.import_metrics(entity)

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py [hunk 16]
```diff
@@ -88,33 +60,35 @@ def db_connect(self):
...
             csr.execute("INSERT INTO hits(id, wikidata_id, wikipedia_hits, europeana_enrichment_hits, europeana_string_hits, pagerank) VALUES (?, ?, ?, ?, ?, ?)", (uri, wikidata_id, wikipedia_hits, europeana_enrichment_hits, europeana_string_hits, pagerank))
             self.db.commit()
-        metrics = {
-            self.METRIC_WIKI_HITS : wikipedia_hits,
-            self.METRIC_ENRICHMENT_HITS : europeana_enrichment_hits,
-            self.METRIC_TERM_HITS : europeana_string_hits,
-            self.METRIC_PAGERANK : pagerank
-        }
-        return metrics
+        return MetricsRecord(uri, 'fake label', wikidata_id, wikipedia_hits, europeana_string_hits, europeana_enrichment_hits, pagerank)
```
Comment (gsergiu) [inline entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py]: please use the extract_def_label method from MetricsImporter to extract the default label. The MetricsImporter.extract_def_label can be updated to take entity as param and to get read of the edm_country

**Review by gsergiu — CHANGES_REQUESTED** [review]
please implement the requested changes

File: entity_collection/munge/mongo_import/config/harvester.properties.template (hunk 1 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/config/harvester.properties.template]: this property was duplicated - removed

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py (hunk 2 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: refactored as required

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py (hunk 3 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: initialization shifted to Harvester subclasses

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py (hunk 4 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: renamed

File: entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py (hunk 5 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ContextClassHarvesters.py]: refactored as required

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py (hunk 12 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py]: refactored as required

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py (hunk 13 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py]: entity object used as parameter

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py (hunk 14 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py:68]: refactored as required

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py (hunk 15 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py:75]: block replaced

File: entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py (hunk 16 above)
Comment (rgraf) [reply entity_collection/munge/mongo_import/entities/ranking_metrics/RelevanceCounter.py]: refactored as required

**Pull request merged**